tensorly/decomposition/_parafac2.py

```diff
@@ -553,6 +672,7 @@ def parafac_updates(X, w, f):
                 (weights, factors, projections),
                 norm_tensor,
                 projected_tensor,
+                mask,
```

Comment:
It's not obvious to me that this is correct. The code is reusing the projected tensor (`P_i @ X_i`) computed on line 641. However, on line 659, the missing values of `X_i` are imputed. I think that means that you need to "re-project" the tensors before you compute the loss if `mask is not None`?

Reply:
Actually @MarieRoald made me realize that the P2 error computation function with mask is probably wrong. 
For now it works as follows:
1. compute the l2 norm of the tensor, at only non-masked values. This works fine.
2. Use the precomputed quantities to compute inner products <T, decomp> and the l2 norm of the model.
This second step is bugged as @MarieRoald mentioned because the projection quantities are outdated, but furthermore, the loss should account for the mask in step 2 as well (the data tensor and the model are compared only at 1-mask values).

The rationale behind steps 1 and 2 is the identity $\|a - b\|_2^2 = \|a\|_2^2 + \|b\|_2^2 - 2<a,b> $, but in the presence of missing data we are computing $\|w*(a-b)\|_2^2\| = \|w*a\|_2^2 + \|w*b\|_2^2 - 2<w*a,w*b>$. I think right now the code computes $\|w*a\|_2^2 + \|b\|_2^2 - 2<w*a_old, b>$ with outdated $a_old$ in the scalar product.

For comparison, in the CP error calculation function, we simply compute the loss naively in the presence of missing data. Because we need to recompute quantities such as the projections before computing the error, it might be similarly effective to just compute the batch cost?

Reply:
Thank you both for the detailed feedback.
@MarieRoald I think you are right as well, I addressed this on b0edd5d6465a6219a7182c1254a33092bf395d65.
@cohenjer I will experiment with both implementations in terms of efficiently and come back at you.

Reply:
@cohenjer I created this small test to investigate the performance, while the output plot is shown below. **Approach (1)** is the existing one that includes recomputing the projections while **approach (2)** "naively" computes the error. There was nothing printed from the `isclose` clause. I guess this means we should  use the latter in the case of missing data? Kindly let me know what you think.
```
def create_random_data(shape, rank):
    I, J, K = shape  # noqa: E741
    pf2 = tl.random.random_parafac2(
        [(J, K) for i in range(I)], rank=rank
    )
    return pf2

import time
import math

tolerance = 1e-8

times1 = []
times2 = []

for i in range(1000000):

    decomposition = create_random_data((3, 80, 25), 3)
    tensor_slices = tl.tensor([np.random.rand(80, 25) for i in range(3)])
    mask = tl.tensor([np.random.randint(0, 2, (80, 25)) for i in range(3)])

    start1 = time.time()

    error1 = _parafac2_reconstruction_error_v1(tensor_slices=tensor_slices, decomposition=decomposition, mask=mask)

    end1 = time.time()

    times1.append(end1 - start1)

    start2 = time.time()

    error2 = _parafac2_reconstruction_error_v2(tensor_slices=tensor_slices, decomposition=decomposition, mask=mask)

    end2 = time.time()

    times2.append(end2 - start2)

    if not math.isclose(error1, error2, abs_tol=tolerance):
        print(f'Error {error1} != {error2} within tolerance {tolerance}')

# do a boxplot of each times1 and times2
plt.boxplot([times1, times2])
plt.yscale('log')
plt.show()
```
![image](https://github.com/user-attachments/assets/dcd66af3-3b9d-4ed9-bb4e-d6376aa2d24f)

For completeness, here is what **approach (2)** looks like:
```
def _parafac2_reconstruction_error_v2(
    tensor_slices, decomposition, norm_matrices=None, projected_tensor=None, mask=None
):
    reconstructed_tensor = parafac2_to_tensor(decomposition)
    total_error = 0
    for i, (slice, slice_mask) in enumerate(zip(tensor_slices, mask)):
        total_error += tl.norm(slice_mask * slice - slice_mask * reconstructed_tensor[i])**2
    return tl.sqrt(total_error)
```

Reply:
Thanks for the precise testing @cchatzis. I assume version 1 is the one you had implemented already in the PR? I think your experiments prove that v1 and the naive error v2 are not the same, and since v2 is obviously the correct loss, this confirms that (1) was not correct. 

Even if it is slower, you should use the v2 error computation in the PR.

Reply:
@cohenjer thank you for your input! v1 refers to the "corrected" version of the one currently in the PR (its not pushed yet). You and @MarieRoald correctly mentioned the need for re-computations and this refers to this "adjusted" version that includes these.

Should I then discard all pre-computations implemented and change the current code so it only computes the error like this (naively) or have an if clause that checks that and if there is no `mask` given utilize the precomputations?

Reply:
You cannot use pre-computations because of the mask, as I explained earlier. If there is not mask then the precomputations, corrected (as in v1) should be fine.

Reply:
@cohenjer addressed this in the latest coments, let me know what you think :)

Reply:
Really strong stuff @cchatzis, great job testing the two methods 💪

Reply:
looks good to me as well !

Reply:
Great comparisons and investigation, thank you!